src/compiler/checker.ts

```diff
@@ -15643,34 +15643,35 @@ namespace ts {
             return getInferredTypes(context);
         }
 
-        function checkTypeArguments(signature: Signature, typeArgumentNodes: ReadonlyArray<TypeNode>, typeArgumentTypes: Type[], reportErrors: boolean, headMessage?: DiagnosticMessage): boolean {
+        function checkTypeArguments(signature: Signature, typeArguments: ReadonlyArray<TypeNode>, reportErrors: boolean, headMessage?: DiagnosticMessage): Type[] | false {
+            const isJavascript = isInJavaScriptFile(signature.declaration);
             const typeParameters = signature.typeParameters;
-            let typeArgumentsAreAssignable = true;
+            const typeArgumentTypes = fillMissingTypeArguments(map(typeArguments, getTypeFromTypeNode), typeParameters, getMinTypeArgumentCount(typeParameters), isJavascript);
             let mapper: TypeMapper;
-            for (let i = 0; i < typeArgumentNodes.length; i++) {
-                if (typeArgumentsAreAssignable /* so far */) {
-                    const constraint = getConstraintOfTypeParameter(typeParameters[i]);
-                    if (constraint) {
-                        let errorInfo: DiagnosticMessageChain;
-                        let typeArgumentHeadMessage = Diagnostics.Type_0_does_not_satisfy_the_constraint_1;
-                        if (reportErrors && headMessage) {
-                            errorInfo = chainDiagnosticMessages(errorInfo, typeArgumentHeadMessage);
-                            typeArgumentHeadMessage = headMessage;
-                        }
-                        if (!mapper) {
-                            mapper = createTypeMapper(typeParameters, typeArgumentTypes);
-                        }
-                        const typeArgument = typeArgumentTypes[i];
-                        typeArgumentsAreAssignable = checkTypeAssignableTo(
-                            typeArgument,
-                            getTypeWithThisArgument(instantiateType(constraint, mapper), typeArgument),
-                            reportErrors ? typeArgumentNodes[i] : undefined,
-                            typeArgumentHeadMessage,
-                            errorInfo);
+            for (let i = 0; i < typeArguments.length; i++) {
+                const constraint = getConstraintOfTypeParameter(typeParameters[i]);
+                if (constraint) {
+                    let errorInfo: DiagnosticMessageChain;
+                    let typeArgumentHeadMessage = Diagnostics.Type_0_does_not_satisfy_the_constraint_1;
+                    if (reportErrors && headMessage) {

```

Comment:
Maybe you should just take `headMessage` but not `reportErrors` and have the caller pass in a message if they want errors?
`errorInfo` will definitely be undefined in the call to `chain`, so it's not really a chain...

Reply:
Good idea. 

I couldn't find another way to *start* a message chain, and `checkTypeAssignableTo` wants a message chain. But I did use constants as arguments to `chainDiagnosticMessages`, and removed some mutation.

Reply:
Actually, `reportErrors` turns out to be required because sometimes you want to report errors even without a different `headMessage`.

Reply:
Do you know why `checkTypeRelatedTo` takes both `headMessage` and `containingMessageChain`?

Reply:
I took a look at `checkTypeRelatedTo`. The intent is to pass a chain
`containingMessageChain=[e1, e2, e3]`, `headMessage=e6` such that, if `checkTypeRelatedTo` finds errors `e4` and `e5`, the resulting chain is `[e1, e2, e3, e4, e5, e6]`.

Only three uses exist as far as I can tell.
* checkTypeArguments -- singleton 'chain'
* checkTypePredicate -- singleton 'chain', but no headMessage
* checkDecorator -- singleton 'chain', headMessage *might* not be defined. I don't know whether `getDiagnosticHeadMessageForDecoratorResolution` is effectively total.

So the split functionality is used, but the requirement that containingMessageChain actually be a *chain* just makes writing `checkTypeRelatedTo` slightly simpler.